src/simd_128_u16.h

```diff
@@ -0,0 +1,269 @@
+/*
+ * Copyright 2017-2018 Scality
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __NTTEC_SIMD_128_U16_H__
+#define __NTTEC_SIMD_128_U16_H__
+
+#include <x86intrin.h>
+
+namespace nttec {
+
+/** The namespace simd contains functions that are accelerated by using SIMD
+ *  operations over 128bits
+ *
+ *  It supports operations on 16-bit numbers
+ */
+namespace simd {
+namespace u16 {
+
+typedef __m128i am128i __attribute__((aligned(ALIGN_SIZE)));
+
+/* ==================== Essential Operations =================== */
+
+const aint16 F3 = 257;
```

Comment:
I still think namespace are not the good way to do that.

Even without a proper SIMD wrapper (which we should write at some point), you can keep everything in the same namespace (`nttec::simd`), thanks to function overloading.

I mean, it's not a problem to define both:

```c++

inline aint128 mod_after_add(am128i a, aint32 card);
inline aint128 mod_after_add(am128i a, aint16 card);
```

They are different functions (because of their type signatures) and can have different body (using 16-bit and 32-bit element).
Did you tried? I think this could work until we get a proper wrapper.

Reply:
yeah, I will update it. I agree that it's better :)